inject/src/main/java/io/micronaut/context/beans/BeanDefinitionProvider.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2017-2025 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.context.beans;
+
+import io.micronaut.context.BeanContext;
+import io.micronaut.context.BeanResolutionContext;
+import io.micronaut.context.DisabledBean;
+import io.micronaut.context.RuntimeBeanDefinition;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.type.Argument;
+import io.micronaut.inject.BeanConfiguration;
+import io.micronaut.inject.BeanDefinition;
+import io.micronaut.inject.BeanDefinitionReference;
+import io.micronaut.inject.QualifiedBeanType;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+/**
+ * The API to access bean definitions and bean references.
+ * Internal implementation is responsible for loading bean definitions and checking if it's enabled.
+ *
+ * @author Denis Stepanov
+ * @since 5.0
+ */
+public interface BeanDefinitionProvider {
```

Comment:
why do we need a new public interface? The name is also inconsistent with `BeanProvider` and may cause confusion. It seems more like a service or registry type object. Since `BeanDefinitionRegistry` is already taken it should be something like `BeanDefinitionService`.

Also does it need to be public? Duplicating public APIs will cause confusing IMO

Reply:
It also seems to reference internal implementation details. Would make this non-public, sealed and annotated with `@Internal`